### PR TITLE
Feature hybrid to composition

### PR DESCRIPTION
**Pull request by MeineZ**
The deferred composition task has been added to the hybrid frame graph. It now uses the reflection and shadow output from the ray tracing task to render the scene with reflections and shadows. The full deferred frame graph samples its reflections from the skybox.

**Has to be tested on an RTX card, as the ray tracing code has changed** (will update with comment)

[Picture showing this working in the hybrid frame graph](https://gyazo.com/937dd26bce8339fc225d5fb474661419.jpg)

### CODE REVIEW

File: demo/scene_viknell.hpp [hunk 1]
```diff
@@ -74,6 +75,8 @@ namespace viknell_scene
 		//pos.m128_f32[0] = sin(t * 0.1) * 0.5;
 		//test_model->SetPosition(pos);
 
+		point_light_node->SetPosition({float(sin(t * 0.1f)) * 0.9f, 0.f, 1.2f});
+
```
Comment (MeineZ) [inline demo/scene_viknell.hpp]: Revert viknell scene

File: resources/shaders/deferred_composition.hlsl [hunk 2]
```diff
@@ -62,11 +65,27 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
...
+			is_hybrid);
+		shadow_factor = clamp(shadow_factor, ambient, 1.0);
 		
-		float3 skybox_reflection = skybox.SampleLevel(s0, SampleSphericalMap(reflect(-V, normal)), 0);
+		// Get reflection
+		float3 reflection = lerp(
+			// Sample from skybox if its hybrid rendering
+			skybox.SampleLevel(s0, SampleSphericalMap(reflect(-V, normal)), 0),
+			// Reflection buffer if its hybrid rendering
+			buffer_refl_shadow[screen_coord].xyz,
```
Comment (MeineZ) [inline resources/shaders/deferred_composition.hlsl]: Comments here are wrong

File: src/render_tasks/d3d12_deferred_composition.hpp [hunk 3]
```diff
@@ -138,16 +172,23 @@ namespace wr
...
-					data.out_irradiance = static_cast<d3d12::TextureResource*>(pred_data.out_irradiance.m_pool->GetTexture(pred_data.out_irradiance.m_id));
+					data.out_irradiance = static_cast<d3d12::TextureResource*>(scene_graph.GetCurrentSkybox()->m_irradiance->m_pool->GetTexture(scene_graph.GetCurrentSkybox()->m_irradiance->m_id));
 					d3d12::CreateSRVFromTexture(data.out_irradiance);
 				}
 
+				// We need a fallback texture when the composition task is not executed in the hybrid
+				if(scene_graph.m_skybox.has_value() && !data.is_hybrid)
+				{
+					// Set skybox as hybrid fallback?
+				}
```
Comment (MeineZ) [inline src/render_tasks/d3d12_deferred_composition.hpp]: Remove this unnecessary if

File: src/render_tasks/d3d12_deferred_composition.hpp [hunk 4]
```diff
@@ -212,7 +253,8 @@ namespace wr
 
 	}
 
-	inline void AddDeferredCompositionTask(FrameGraph& fg, std::optional<unsigned int> target_width, std::optional<unsigned int> target_height)
+	template<typename T = DeferredCompositionTaskData>
+	inline void AddDeferredCompositionTask(FrameGraph& fg, std::optional<unsigned int> target_width, std::optional<unsigned int> target_height, bool is_in_hybrid = false)
```
Comment (MeineZ) [inline src/render_tasks/d3d12_deferred_composition.hpp]: Remove unnecessary parameter 'is_in_hybrid'

File: resources/shaders/rt_global.hlsl [hunk 5]
```diff
@@ -3,8 +3,16 @@
 //#define PATH_TRACING
 //#define DEPTH_OF_FIELD
 #define EPSILON 0.01
-//#define SOFT_SHADOWS
+#define SOFT_SHADOWS
 #define MAX_SHADOW_SAMPLES 4
 
+#define FALLBACK
```
Comment (VZout) [inline resources/shaders/rt_global.hlsl]: Don't like this but I'll do it better in the future. its 5 minutes of work.

**Pull request merged**